File: D-MergeIntervals/cpp/problem1.cpp

#include<bits/stdc++.h>
using namespace std;

vector<vector<int>> merge(vector<vector<int>>& intervals) {
    if (intervals.empty()) return {};

    // Sort the intervals based on their starting points
    sort(intervals.begin(), intervals.end());

    // Initialize the result array
    vector<vector<int>> result;
    result.push_back(intervals[0]);

    // Iterate through the sorted intervals and merge them
    for (int i = 1; i < intervals.size(); ++i) {
        vector<int>& current_interval = intervals[i];
        vector<int>& last_interval = result.back();

        if (current_interval[0] <= last_interval[1]) {
            // Merge the current interval with the last interval
            last_interval[1] = max(last_interval[1], current_interval[1]);
        } else {
            // Add the current interval to the result
            result.push_back(current_interval);
        }
    }

    return result;
}

int main() {
    vector<vector<int>> intervals1 = {{1, 3}, {2, 6}, {8, 10}, {15, 18}};
    vector<vector<int>> result1 = merge(intervals1);
    for (const auto& interval : result1) {
        cout << "[" << interval[0] << "," << interval[1] << "] ";
    }
    cout << endl;

    vector<vector<int>> intervals2 = {{1, 4}, {4, 5}};
    vector<vector<int>> result2 = merge(intervals2);
    for (const auto& interval : result2) {
        cout << "[" << interval[0] << "," << interval[1] << "] ";
    }
    cout << endl;

    vector<vector<int>> intervals3 = {{1, 4}, {0, 4}};
    vector<vector<int>> result3 = merge(intervals3);
    for (const auto& interval : result3) {
        cout << "[" << interval[0] << "," << interval[1] << "] ";
    }
    cout << endl;

    return 0;
}

File: D-MergeIntervals/ruby/problem1.rb
=begin
Merge Intervals
This approach efficiently merges intervals using sorting + iteration.

Approach: Sorting + Merging
Steps:
  Sort the Intervals
    Sort the given intervals based on their starting points (first element of each sub-array).
    Sorting ensures that overlapping intervals are adjacent.

  Iterate and Merge Intervals
    Initialize an empty result array.
    Traverse through sorted intervals and:
      If the current interval overlaps with the last interval in the result, merge them.
      Otherwise, add the current interval to the result as a separate interval.
=end


def merge(intervals)
  return [] if intervals.empty?
  intervals.sort!
  result = [intervals[0]]
  for i in 1...intervals.length
    current_interval = intervals[i]
    last_interval = result[-1]

    if current_interval[0] <= last_interval[1]
      last_interval[1] = [last_interval[1], current_interval[1]].max
    else
      result << current_interval
    end
  end
  return result
end


# optimized version
def merge(intervals)
  return [] if intervals.empty?
  intervals.sort_by! { |interval| interval[0] }
  result = []
  intervals.each do |current_interval|
    if result.empty? || result[-1][1] < current_interval[0]
      result << current_interval
    else
      result[-1][1] = [result[-1][1], current_interval[1]].max
    end
  end
  result
end


intervals = [[1,3],[2,6],[8,10],[15,18]]
puts merge(intervals).to_s
intervals = [[1,4],[4,5]]
puts merge(intervals).to_s
intervals = [[1,4],[0,4]]
puts merge(intervals).to_s

intervals =  [[1,2],[1,5],[2,6],[7,8],[4,6]]
puts merge(intervals).to_s

# Test cases
  # Input: [[1,3],[2,6],[8,10],[15,18]]
  # Output: [[1,6],[8,10],[15,18]]
  # Explanation: Since intervals [1,3] and [2,6] overlap, merge them into [1,6].
  # Input: [[1,4],[4,5]]
  # Output: [[1,5]]
  # Explanation: Intervals [1,4] and [4,5] are considered overlapping.
  # Input: [[1,4],[0,4]]
  # Output: [[0,4]]
  # Explanation: Intervals [1,4] and [0,4] are considered overlapping.
  #  Complexity Analysis
  # Time complexity: O(nlogn) where n is the number of intervals.
  # Sorting the intervals takes O(nlogn) time.
  # Merging the intervals takes O(n) time.
  # Space complexity: O(n) where n is the number of intervals.
  # The space complexity is O(n) because we need to store the sorted intervals.

# Dry run
# Input: [[1,3],[2,6],[8,10],[15,18]]
# Sorted: [[1,3],[2,6],[8,10],[15,18]]
# Initial result: [[1,3]]
# Iteration 1: current_interval = [2,6], last_interval = [1,3]
# Overlapping, merge: result = [[1,6]]
# Iteration 2: current_interval = [8,10], last_interval = [1,6]
# Not overlapping, add: result = [[1,6], [8,10]]
# Iteration 3: current_interval = [15,18], last_interval = [8,10]
# Not overlapping, add: result = [[1,6], [8,10], [15,18]]
# Final result: [[1,6], [8,10], [15,18]]

# Input: [[1,4],[4,5]]
# Sorted: [[1,4],[4,5]]
# Initial result: [[1,4]]
# Iteration 1: current_interval = [4,5], last_interval = [1,4]
# Overlapping, merge: result = [[1,5]]
# Final result: [[1,5]]

# Input: [[1,4],[0,4]]
# Sorted: [[0,4],[1,4]]
# Initial result: [[0,4]]
# Iteration 1: current_interval = [1,4], last_interval = [0,4]
# Overlapping, merge: result = [[0,4]]
# Final result: [[0,4]]